storage/block/block.go

```diff
@@ -31,95 +33,260 @@ import (
-	return b.start
+	b.RLock()
+	start := b.startWithLock()
+	b.RUnlock()
+	return start
+}
+
+func (b *dbBlock) startWithLock() time.Time {
+	return time.Unix(0, b.startUnixNano)
+}
+
+func (b *dbBlock) LastAccessTime() time.Time {
+	return time.Unix(0, atomic.LoadInt64(&b.accessedUnixNano))
+}
+
+func (b *dbBlock) Len() int {
+	b.RLock()
+	length := b.length
+	b.RUnlock()
+	return length
+}
+
+func (b *dbBlock) Checksum() uint32 {
+	b.RLock()
+	checksum := b.checksum
+	b.RUnlock()
+	return checksum
 }
 
-func (b *dbBlock) Checksum() *uint32 {
-	cksum := b.checksum
-	return &cksum
+func (b *dbBlock) OnRetrieveBlock(
+	id ts.ID,
+	startTime time.Time,
+	segment ts.Segment,
+) {
+	b.Lock()
+	defer b.Unlock()
+
+	if b.closed ||
+		!id.Equal(b.retrieveID) ||
+		!startTime.Equal(b.startWithLock()) {
+		return
+	}
+
+	b.resetSegmentWithLock(segment)
 }
 
 func (b *dbBlock) Stream(blocker context.Context) (xio.SegmentReader, error) {
+	b.RLock()
+
 	if b.closed {
+		b.RUnlock()
 		return nil, errReadFromClosedBlock
 	}
+
 	b.ctx.DependsOn(blocker)
-	// If the block is not writable, and the segment is empty, it means
-	// there are no data encoded in this block, so we return a nil reader.
-	if b.segment.Head == nil && b.segment.Tail == nil {
-		return nil, nil
+
+	// Use an int64 to avoid needing a write lock for
+	// the high frequency stream method
+	atomic.StoreInt64(&b.accessedUnixNano, b.now().UnixNano())
+
+	// If the block retrieve ID is set then it must be retrieved
+	var (
+		stream xio.SegmentReader
+		err    error
+	)
+	if b.retriever != nil {
+		stream, err = b.retriever.Stream(b.retrieveID, b.startWithLock(), b)
+		if err != nil {
+			b.RUnlock()
+			return nil, err
+		}
+	} else {
+		stream = b.opts.SegmentReaderPool().Get()
+		// NB(r): We explicitly create a new segment to ensure references
+		// are taken to the bytes refs and to not finalize the bytes.
+		stream.Reset(ts.NewSegment(b.segment.Head, b.segment.Tail, ts.FinalizeNone))
 	}
-	s := b.opts.SegmentReaderPool().Get()
-	// NB(r): We explicitly create a new segment to ensure references
-	// are taken to the bytes refs and to not finalize the bytes.
-	s.Reset(ts.NewSegment(b.segment.Head, b.segment.Tail, ts.FinalizeNone))
-	blocker.RegisterFinalizer(context.FinalizerFn(s.Close))
-	return s, nil
+
+	if b.mergeTarget != nil {
+		var mergeStream xio.SegmentReader
+		mergeStream, err = b.mergeTarget.Stream(blocker)
+		if err != nil {
+			stream.Finalize()
+			return nil, err
+		}
+		// Return a lazily merged stream
+		// TODO(r): once merged reset this block with the contents of it
+		stream = newDatabaseMergedBlockReader(b.startWithLock(), stream, mergeStream, b.opts)
+	}
+
+	// Register the finalizer for the stream
+	blocker.RegisterFinalizer(stream)
+	b.RUnlock()
+
+	return stream, nil
+}
+
+func (b *dbBlock) IsRetrieved() bool {
+	b.RLock()
+	retrieved := b.retriever == nil
+	b.RUnlock()
+	return retrieved
+}
+
+func (b *dbBlock) Merge(other DatabaseBlock) {
```

Comment:
We use this when a bunch of bootstrapped blocks or drained blocks are added to a series (either with `bufferDrained(...)` or `Bootstrap(...)`) and there is existing blocks.